### PR TITLE
239: Migration from Maven to Gradle for Prime Router

### PR DESCRIPTION
**Pull request by carlosfelix2**
This PR migrates from Maven to Gradle.  I will need to understand the deployment process to make sure everything in the build works as expected.  Run the build using the following commands:
`./gradlew clean` - to delete the old build artifacts
`./gradlew package` - to build everything and run the unit tests
`./quick-test all` - run the quick tests
Run the Docker container
`./gradlew testEnd2End` - To run the end to end tests from the primeCLI

## Changes
- Migrated from Maven to Gradle
- Added Gradle wrapper to eliminate the need to install Gradle separately
- Run primeCLI from Gradle including the end to end tests
- Updated documents
- No longer need a fat jar to run the primeCLI since it now uses Gradle
- Fixed the SFTP Docker container not setting write permissions to the upload folder

## Questions
- Is the prime script still needed?  We can run the CLI from Gradle.

### Testing
- [x] Local build completes successfully including unit tests
- [x] Github Actions build is successful
- [x] Tested locally?
- [x] Ran `quickTest all`?
- [x] Ran end to end tests against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [ ] Added tests?

### Security
- [ ] Did you check for sensitive data, and remove any? 
- [ ] Does logging contain sensitive data?  
- [ ] Are additional approvals needed for this change?
- [ ] Are there potential vulnerabilities or licensing issues with any new dependencies introduced?

### Process
- [x] Includes a summary of what a code reviewer should verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Security
*List potential security threats and mitigations in the PR*

## Fixes
*List GitHub issues this PR fixes*
- [Issue #239](https://github.com/CDCgov/prime-data-hub/issues/239)

## To Be Done
- Need others to test the build to make sure there are no OS specific issues.

### CODE REVIEW

File: prime-router/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    }
+    implementation("com.azure:azure-identity:1.2.5") {
+        exclude(group = "com.azure", module = "azure-core")
+        exclude(group = "com.azure", module = "azure-core-http-netty")
+    }
+    implementation("org.apache.logging.log4j:log4j-api:[2.13.2,)")
+    implementation("org.apache.logging.log4j:log4j-core:[2.13.2,)")
+    implementation("org.apache.logging.log4j:log4j-slf4j-impl:[2.13.2,)")
+    implementation("org.apache.logging.log4j:log4j-api-kotlin:1.0.0")
+    implementation("com.github.doyaaaaaken:kotlin-csv-jvm:0.15.1")
```
Comment (RickHawesUSDS) [inline prime-router/build.gradle.kts:266]: Upgrade to 0.15.2.

**Review by RickHawesUSDS — CHANGES_REQUESTED** [review]
Super excited by this improvement. Thanks a bunch.

Just a few simple changes requested.

File: prime-router/start_func.sh [hunk 2]
```diff
@@ -15,23 +15,12 @@ base_name=azure-functions/prime-data-hub-router
 # find the function folder
 if [ -d /prime-data-hub-router/$base_name ]; then
   function_folder=/prime-data-hub-router/$base_name
-elif [ -d target/$base_name ]; then
+elif [ -d build/$base_name ]; then
   function_folder=target/$base_name
 elif [ -d $base_name ]; then
```
Comment (RickHawesUSDS) [inline prime-router/start_func.sh:20]: I think this line `function_folder=target/$base_name` should be `function_folder=build/$base_name`

File: prime-router/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+}
+
+val azureScriptsTmpDir = File(rootProject.buildDir.path, "${azureFunctionsDir}-scripts/${azureAppName}")
+val azureScriptsFinalDir = rootProject.buildDir
+tasks.register<Copy>("gatherAzureScripts") {
+    from("./")
+    into(azureScriptsTmpDir)
+    include("prime")
+    include("start_func.sh")
+}
```
Comment (RickHawesUSDS) [inline prime-router/build.gradle.kts:148]: If you notice, the POM does a chmod as well as copies these files. 
```
            <plugin>
                <artifactId>maven-antrun-plugin</artifactId>
                <version>3.0.0</version>
                <executions>
                    <execution>
                        <phase>package</phase>
                        <configuration>
                            <target>
                                <copy file = "prime" tofile = "target/prime"/>
                                <chmod file = "target/prime" perm = "+x"/>
                                <copy file = "start_func.sh" tofile = "target/start_func.sh"/>
                                <chmod file = "target/start_func.sh" perm = "+x"/>
                            </target>
                        </configuration>
                        <goals>
                            <goal>run</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```
Marking these files execute is needed to make `docker-compose up` work.

**Comment by MauriceReeves-usds** [comment]
@carlosfelix2 - if you've addressed Rick's comments above, go ahead and resolve those conversations, and I'll take a look as well.

File: prime-router/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+val dbPassword = (project.properties["DB_PASSWORD"] ?: "changeIT!") as String
+val dbUrl = (project.properties["DB_URL"] ?: "jdbc:postgresql://localhost:5432/prime_data_hub") as String
+val jooqSourceDir = "build/generated-src/jooq/src/main/java"
+val jooqPackageName = "gov.cdc.prime.router.azure.db"
+
+defaultTasks("package")
+
+// Set the compiler JVM target
+java {
+    sourceCompatibility = JavaVersion.VERSION_1_8
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/build.gradle.kts:43]: Happy to hear!  Updated.

File: prime-router/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    targetCompatibility = JavaVersion.VERSION_1_8
+}
+val compileKotlin: KotlinCompile by tasks
+val compileTestKotlin: KotlinCompile by tasks
+compileKotlin.kotlinOptions.jvmTarget = "1.8"
+compileTestKotlin.kotlinOptions.jvmTarget = "1.8"
+
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/build.gradle.kts:50]: Updated.

File: prime-router/build.gradle.kts (hunk 1 above)
Comment (carlosfelix2) [reply prime-router/build.gradle.kts:266]: Updated.

File: prime-router/start_func.sh (hunk 2 above)
Comment (carlosfelix2) [reply prime-router/start_func.sh:20]: Yes, I must have changed it at some point.  Thx.

File: prime-router/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    dependsOn("copyAzureResources")
+    dependsOn("copyAzureScripts")
+}
+
+repositories {
+    mavenCentral()
+    maven {
+        url = uri("https://kotlin.bintray.com/kotlinx")
+    }
+    maven {
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/build.gradle.kts:234]: In order to remove that repo, I had to force the use of org.jetbrains.kotlinx:kotlinx-coroutines-core that is required by com.github.KennethWussmann:mock-fuel:1.3.0.  mock-fuel uses 1.3.3-native-mt which is no longer available in the Maven repo.  I overwrote that to use 1.3.5-native-mt from the Maven repo.  The tests run fine.

File: prime-router/build.gradle.kts (hunk 3 above)
Comment (carlosfelix2) [reply prime-router/build.gradle.kts:148]: Done.  I can see my changes took effect, but I am unable to fully test in Windows as I never has an issue.  Please pull and retest doing a full build:
`./gradlew clean package`

File: prime-router/build.gradle.kts (hunk 1 above)
Comment (carlosfelix2) [reply prime-router/build.gradle.kts:266]: Will need to keep track of when https://github.com/CDCgov/prime-data-hub/pull/941 gets merged in.

**Comment by carlosfelix2** [comment]
> @carlosfelix2 - if you've addressed Rick's comments above, go ahead and resolve those conversations, and I'll take a look as well.

Changes made and ready for your test

File: prime-router/build.gradle.kts (hunk 3 above)
Comment (carlosfelix2) [reply prime-router/build.gradle.kts:151]: @RickHawesUSDS Arg!  I am changing my approach.  Could you pull and try again?

Comment (RickHawesUSDS) [reply prime-router/build.gradle.kts:151]: I see the problem. Change the code to set the file in `azureScriptsFinalDir`... 
```
        File(azureScriptsFinalDir.path, primeScriptName).setExecutable(true);
        File(azureScriptsFinalDir.path, startFuncScriptName).setExecutable(true);
```

Comment (RickHawesUSDS) [reply prime-router/build.gradle.kts:151]: With this change, I can run the functions locally on my mac.

Comment (RickHawesUSDS) [reply prime-router/build.gradle.kts:151]: BTW: The pulled the latest and that doesn't work. I think your latest changes didn't fix the problem, but your previous approach did work with my change.

Comment (carlosfelix2) [reply prime-router/build.gradle.kts:151]: Thank you.  I really wanted to find a way that I could truly verify in Windows, but seems that is not going to be the case.  I'll use the approach you took.  :(

Comment (RickHawesUSDS) [reply prime-router/build.gradle.kts:158]: Current branch is working for me.

**Review by RickHawesUSDS — APPROVED** [review]

**Comment by RickHawesUSDS** [comment]
Looking good. Thanks, Carlos.

**Pull request merged**